private/buf/bufgen/migrate.go

```diff
@@ -0,0 +1,553 @@
+		ctx,
+		logger,
+		input,
+		types,
+		includePaths,
+		excludePaths,
+	)
+	if err != nil {
+		return nil, err
+	}
+	externalConfigV2.Inputs = []bufgenv2.ExternalInputConfigV2{
+		*inputConfig,
+	}
+	for _, pluginConfigV1 := range configV1.PluginConfigs {
+		pluginConfigV2, err := pluginConfigToExternalPluginConfigV2(
+			pluginConfigV1,
+			findPluginFunc,
+			includeImports,
+			includeWKT,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("unable to migrate plugin %q: %w", pluginConfigV1.PluginName(), err)
+		}
+		externalConfigV2.Plugins = append(externalConfigV2.Plugins, *pluginConfigV2)
+	}
+	externalConfigV2.Managed = *managedConfigV1ToExternalManagedConfigV2(configV1.ManagedConfig)
+	return &externalConfigV2, nil
+}
+
+func pluginConfigToExternalPluginConfigV2(
+	pluginConfig bufgenplugin.PluginConfig,
+	findPluginFunc func(string) (string, error),
+	includeImports bool,
+	includeWKT bool,
+) (*bufgenv2.ExternalPluginConfigV2, error) {
+	externalPluginConfig := bufgenv2.ExternalPluginConfigV2{}
+	// opt and out are common to all plugins
+	externalPluginConfig.Out = pluginConfig.Out()
+	optString := pluginConfig.Opt()
+	switch opts := strings.Split(optString, ","); len(opts) {
+	case 1:
+		if optString == "" {
+			externalPluginConfig.Opt = nil
+			break
+		}
+		externalPluginConfig.Opt = opts[0]
+	default:
+		externalPluginConfig.Opt = opts
+	}
+	pluginName := pluginConfig.PluginName()
+	switch t := pluginConfig.(type) {
+	case bufgenplugin.BinaryPluginConfig:
+		strategy := t.Strategy().String()
+		externalPluginConfig.Strategy = &strategy
+		externalPluginConfig.Binary = t.Path()
+		if len(t.Path()) == 1 {
+			externalPluginConfig.Binary = t.Path()[0]
+		}
+	case bufgenplugin.ProtocBuiltinPluginConfig:
+		strategy := t.Strategy().String()
+		externalPluginConfig.Strategy = &strategy
+		externalPluginConfig.ProtocBuiltin = &pluginName
+		if protocPath := t.ProtocPath(); protocPath != "" {
+			externalPluginConfig.ProtocPath = &protocPath
+		}
+	case bufgenplugin.LocalPluginConfig:
+		strategy := t.Strategy().String()
+		externalPluginConfig.Strategy = &strategy
+		if binaryPath, err := findPluginFunc("protoc-gen-" + pluginName); err == nil {
+			// this is a binary plugin
+			externalPluginConfig.Binary = binaryPath
+			break
+		}
+		if _, isProtocBuiltin := bufpluginexec.ProtocProxyPluginNames[pluginName]; isProtocBuiltin {
+			externalPluginConfig.ProtocBuiltin = &pluginName
+			break
+		}
+		// At this point, we know for certain that this plugin is not protoc-builtin.
+		// It's possible that the plugin is a valid binary plugin but not installed on
+		// the user's environment.
+		// TODO: we can also treat it as a binary plugin, leaving a comment like so:
+		// # we couldn't find protoc-gen-xyz locally and you should verify that it is
+		// # a binary installed locally
+		// binary_plugin: protoc-gen-xyz
+		return nil, fmt.Errorf("plugin protoc-gen-%s is not found locally and %s is not built-in to protoc", pluginName, pluginName)
```

Comment:
Checking from `PATH` and then checking from the list of protoc plugins is how we have been deciding the type of plugin.